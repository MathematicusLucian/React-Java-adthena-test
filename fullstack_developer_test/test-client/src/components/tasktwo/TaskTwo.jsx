import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import dao from '@services/dao';

import {ThemeProvider} from "styled-components";
import { GlobalStyles } from "../globalStyles";
import { lightTheme, darkTheme } from "../themes";

// This is a functional component

const TaskTwo = ({ message }) => {

  const userApiRequest = url => {
  
    const [data, setData] = useState([]);
    const [isLoaded, setIsLoaded] = useState(false);
    const [error, setError] = useState(null);
  
    useEffect(() => {
      dao.GET(`${url}`)
      .then((res) => {
        setData(res);
        setIsLoaded(true);
      })
      .catch(() => {
        setHasError(error);
      });
    }, [url]);
  
    return { error, isLoaded, data };
  };

  const [theme, setTheme] = useState('light');
  const themeToggler = () => {
    theme === 'light' ? setTheme('dark') : setTheme('light')
  }

  const [User, setUser] = useState('');

  const handleChange = (e) => {
    setUser(e.target.value);
  }

  const { data, error, isLoaded } = userApiRequest(
    "http://127.0.0.1:8082/api/todos/" + `${User}`
  );

  if (error !== null) {
    return <div>Error: {error.message}</div>;
  }
  
  return (

    <ThemeProvider theme={theme === 'light' ? lightTheme : darkTheme}>
      <>
      <GlobalStyles/>

      <div className="task">
        
        <button id="toggle" onClick={themeToggler}>Switch Theme</button>

        <h1>Task Two</h1>
        <div className="content">

            <input
                type="text"
                placeholder="Type in a user, e.g. Bret"
                value={User.trim()} 
                onChange={handleChange}
            />

            <h2>User Profile</h2>

            <h3 className="userRequested">
              {User.trim()}
            </h3>

            {error && (<span>| Server error...</span>)}
            {!isLoaded && (<span>| Loading...</span>)}
            {(data.code === 404 || (data.length < 1)) && (
              <span>User not found</span>
            )}

            {!data.code && data.length >= 1 && (
              <>
                <div>
                  {data.map((item) => (
                    <div key={item.id}>
                      <ul>
                        <li><strong>User ID:</strong> {item.id}</li>
                        <li><strong>Email:</strong> {item.email}</li>
                        <li><strong>Website:</strong> {item.website}</li>
                      </ul>
                      <h3>Todo List</h3>
                      <h4>Complete</h4>
                      {item.todos
                        .filter(t => t.completed)
                        .map((t) => (
                            <div key={t.id}>
                              {t.id}: {t.title}
                            </div>
                      ))}
                      <h4>Incomplete</h4>
                      {item.todos
                        .filter(t => !t.completed)
                        .map((t) => (
                            <div key={t.id}>
                              {t.id}: {t.title}
                            </div>
                      ))}
                      </div>
                  ))}
                </div>
              </>
            )}

          </div>
     </div>

      </>
    </ThemeProvider>

    );
};

TaskTwo.propTypes = {
  message: PropTypes.string,
};

TaskTwo.defaultProps = {
  message: 'Adthena',
};

export default TaskTwo;
