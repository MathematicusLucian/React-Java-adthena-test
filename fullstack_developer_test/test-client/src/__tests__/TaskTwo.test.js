import React from 'react';
import { shallow, render, mount } from "enzyme";
import TaskTwo from './../components/tasktwo/TaskTwo';

describe('TaskTwo', () => {

  let wrapper;
  beforeEach(() => {
    const props = { changeRoute: jest.fn() };
    wrapper = shallow(<TaskTwo {...props}></ TaskTwo>);
  });

  it('should render correctly', () => {
    shallow(<TaskTwo />);
  });

  it('should render - shallow', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('should render (snapshot)', () => {
    const rendered = render(<div className="task"/>)
    expect(rendered).toMatchSnapshot();
  });

  it('should render correctly in "debug" mode', () => {
    const component = shallow(<TaskTwo debug />)
    expect(component).toMatchSnapshot();
  });

  it('task class should exist', () => {;
    expect(wrapper.find(".task").length).toEqual(1);
  });

  it('should render an input tag', () => {
    expect(wrapper.find('input').exists()).toBe(true);
  });

  it('should render an input tag - 2', () => {
    expect(wrapper.find("input").length).toEqual(1);
  });

  it('the default value for the input field should be empty', () => {
    expect(wrapper.find('input').prop('value')).toBe('');
  });

  it('should set the user value on change event with trim', () => {
    wrapper.find('input').simulate('change', {
      target: {
        value: 'John  ',
      },
    });
    expect(wrapper.find('input').prop('value')).toEqual(
      'John',
    );
  });

  it('on change of value in the field, the state of that field in the component should be updated', () => {
    wrapper.find('input').simulate('change', {
      target: {
        value: 'Ivan',
      },
    });
    expect(wrapper.find('input').prop('value')).toBe(
      'Ivan'
    );
    wrapper.find('input').simulate('change', {
      target: {
        value: 'David',
      },
    });
    expect(wrapper.find('input').prop('value')).toBe(
      'David'
    );
  });

  it('on input change, a submit handler function should be triggered on click event', () => {
    wrapper.find('input').simulate('change', {
      target: {
        value: 'Ignatius',
      },
    });
    wrapper = wrapper.update();
    expect(wrapper.find('h3.userRequested').text()).toBe("Ignatius");
  });

  it('on clicking theme switch button, toggle background colour (is light on load)', () => {
    const fn = jest.fn();
    const wrapper = shallow(<TaskTwo onClick={fn} />);
    wrapper.find('ToggleButton').simulate('click');
    expect(wrapper.props().theme.background).toEqual("#999");
    wrapper.find('ToggleButton').simulate('click');
    expect(wrapper.props().theme.background).toEqual("#363537");
    wrapper.find('ToggleButton').simulate('click');
    expect(wrapper.props().theme.background).toEqual("#999");
  });

});