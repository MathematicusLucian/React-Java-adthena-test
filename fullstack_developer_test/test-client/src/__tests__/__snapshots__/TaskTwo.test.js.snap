// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTwo should render (snapshot) 1`] = `
<div
  class="task"
/>
`;

exports[`TaskTwo should render correctly in "debug" mode 1`] = `
<Component
  theme={
    Object {
      "background": "#363537",
      "body": "#FFF",
      "text": "#363537",
      "toggleBorder": "#FFF",
    }
  }
>
  <Memo(l) />
  <div
    className="task"
  >
    <button
      id="toggle"
      onClick={[Function]}
    >
      Switch Theme
    </button>
    <h1>
      Task Two
    </h1>
    <div
      className="content"
    >
      <input
        onChange={[Function]}
        placeholder="Type in a user, e.g. Bret"
        type="text"
        value=""
      />
      <h2>
        User Profile
      </h2>
      <h3
        className="userRequested"
      />
      <span>
        | Loading...
      </span>
      <span>
        User not found
      </span>
    </div>
  </div>
</Component>
`;
